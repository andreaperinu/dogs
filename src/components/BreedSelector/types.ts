import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface IBreedSelectorProps {
	maybeSelectedBreed: O.Option<IBreed>,
	maybeBreeds: O.Option<IBreed[]>,
	onChange: (maybeBreed: O.Option<IBreed>) => void
}
