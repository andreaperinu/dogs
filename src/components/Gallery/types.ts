import * as O from 'fp-ts/lib/Option';

export interface IGalleryProps {
	maybeImages: O.Option<string[]>
}
