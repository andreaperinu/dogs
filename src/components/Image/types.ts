import * as O from 'fp-ts/lib/Option'

export interface IImageProps {
	maybeImage: O.Option<string>
	fixedSize?: boolean
}
