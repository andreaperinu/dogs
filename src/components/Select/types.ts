import * as O from 'fp-ts/lib/Option';
import { BaseComponent } from '../../types';

export interface ISelectProps extends BaseComponent {
	maybeValue: O.Option<ISelectItem>
	maybeOptions: O.Option<ISelectItem[]>
	maybeText?: O.Option<string>
	disabled?: boolean
	onChange: (item: O.Option<ISelectItem>) => void
}

export interface ISelectItem {
	id: string
	value: string
}

export interface ISelectStyledProps {
	showOptions: boolean
	disabled: boolean
}
