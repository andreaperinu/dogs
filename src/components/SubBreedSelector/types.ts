import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface ISubBreedSelectorProps {
	disabled?: boolean
	maybeSelectedSubBreed: O.Option<string>,
	maybeBreed: O.Option<IBreed>,
	onChange: (maybeSubBreed: O.Option<string>) => void
}

export interface ISubBreedSelectorStyledProps {
	disabled: boolean
}
