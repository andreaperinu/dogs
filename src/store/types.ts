import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

import * as A from './actions/rawActions';

type ActionType = keyof typeof A

export interface IStoreState {
	maybeBreeds: O.Option<IBreed[]>,
	maybeRandomImage: O.Option<string>,
	maybeError: O.Option<string>,
	maybeImages: O.Option<string[]>
}

export type StoreAction = (type: ActionType, payload: unknown) => Promise<Partial<IStoreState>>

export type StoreReducer = (state: IStoreState) => Promise<Partial<IStoreState>>
export type TypedStoreReducer<T> = (state: IStoreState, payload: T) => Promise<Partial<IStoreState>>

export type StoreDispatch = ({ type, payload }: { type: ActionType, payload?: unknown }) => void

export type Action<P = undefined> = (payload?: P) => ({ type: ActionType, payload?: P })
