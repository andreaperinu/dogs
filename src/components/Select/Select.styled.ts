import styled, { css } from 'styled-components';

import { ISelectStyledProps } from './types';

export const SelectStyled = styled.div<ISelectStyledProps>`
${({ showOptions, disabled, theme: {
	border, borderRadius, minHeight, brandPrimary
} }) => css`

		position: relative;
		display: flex;
		align-items: center;

		.wrapper {
			width: 100%;
			max-width: 20rem;
			position: relative;
		}

		.list {
			opacity: 0;
			height: 0;
      margin: 0;
			padding: 0;
			transition: opacity .5s;
			overflow: hidden;
			list-style-type: none;
			background: white;
    	max-height: 20rem;
    	overflow: auto;
			
			${showOptions && css`
				border: ${border};
				height: auto;
				opacity: 1;
			`}
		}

		.value {
			border: ${border};
			border-radius: ${borderRadius};
      padding: .5rem;
			min-height: ${minHeight};
			box-sizing: border-box;
			white-space: nowrap;
			overflow: hidden;
			text-overflow: ellipsis;

			${!disabled && css` cursor: pointer; `}
		}

		li {
      padding: .5rem;

			${showOptions && css` cursor: pointer; `}
			&:hover { background: ${brandPrimary}; }
		}

		.list {
			border-top: none;
			position: absolute;
			width: 100%;
			box-sizing: border-box;
			z-index: 1;
		}

		.text {
			${!disabled && css` cursor: pointer; `}
			margin-left: .5rem;
		}

	`}
`
