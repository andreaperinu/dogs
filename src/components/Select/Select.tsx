import classNames from 'classnames';
import * as A from 'fp-ts/lib/Array';
import * as B from 'fp-ts/lib/boolean';
import { constVoid, pipe } from 'fp-ts/lib/function';
import * as O from 'fp-ts/lib/Option';
import React, { memo, useCallback, useMemo } from 'react';

import { noopComponent } from '@/utils/utils';

import { SelectStyled } from './Select.styled';
import { ISelectItem, ISelectProps } from './types';

export const BaseSelect = ({
  maybeValue, maybeOptions, className, onChange, disabled = false, maybeText = O.none,
}: ISelectProps) => {
  const [showOptions, setShowOptions] = React.useState(false)

  const onItemSelect = useCallback((item: ISelectItem) => {
    setShowOptions(false)
    onChange(O.fromNullable(item))
  }, [onChange])

  const onShowOptionsHandler = useCallback(() => pipe(
    !disabled,
    B.fold(constVoid, () => setShowOptions((prev) => !prev))
  ), [disabled])

  const renderedValue = useMemo(() => pipe(
    maybeValue,
    O.map(({ value }) => value),
    O.getOrElse(() => ''),
    v => <div className={classNames('value', className)} onClick={onShowOptionsHandler}>{v}</div>
  ), [maybeValue, onShowOptionsHandler])

  const renderedOptions = useMemo(() => pipe(
    maybeOptions,
    O.map(A.map(option => (
      <li key={option.id} value={option.id} onClick={() => onItemSelect(option)}>
        {option.value}
      </li>
    ))),
    O.map(O.fromPredicate(A.isNonEmpty)),
    O.chain(O.map(items => <ul className="list">{items}</ul>)),
    O.getOrElse(noopComponent),
  ), [maybeOptions, onItemSelect])

  const renderedText = useMemo(() => pipe(
    maybeText,
    O.map(text => <div className="text" onClick={onShowOptionsHandler}>{text}</div>),
    O.getOrElse(noopComponent)
  ), [maybeText, onShowOptionsHandler])

  return (
    <SelectStyled showOptions={showOptions} disabled={disabled}>

      <div className="wrapper">
        {renderedValue}
        {renderedOptions}
      </div>

      {renderedText}
    </SelectStyled>
  )
}

export const Select = memo(BaseSelect)
