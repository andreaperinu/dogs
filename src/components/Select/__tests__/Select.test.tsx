import * as O from 'fp-ts/lib/Option';
import React from 'react';
import { act } from 'react-dom/test-utils';

import { setup } from '@/tests/utils';

import { BaseSelect as component } from '../Select';
import { ISelectProps } from '../types';

describe('<Select />', () => {
  const onChange = jest.fn()
  const useStateMock = jest.fn()

  const originalUseState = React.useState

  const value = { id: '0', value: 'name_0' }

  const props: ISelectProps = {
    maybeValue: O.some(value),
    maybeOptions: O.some([
      value,
      { id: '1', value: 'name_1' },
    ]),
    onChange,
  }


  beforeEach(() => {
    React.useState = originalUseState
    onChange.mockClear()
    useStateMock.mockClear()
  })

  it('should match the snapshot', () => {
    const wrapper = setup({ component, props })
    expect(wrapper).toMatchSnapshot()
  })

  it('should match the snapshot (blank value)', () => {
    const wrapper = setup({ component, props: { ...props, maybeValue: O.none } })
    expect(wrapper).toMatchSnapshot()
  })

  it('should match the snapshot (show text)', () => {
    const wrapper = setup({ component, props: { ...props, maybeText: O.some('text') } })
    expect(wrapper).toMatchSnapshot()
  })

  it('should match the snapshot (show options)', () => {
    const wrapper = setup({ component, props, shouldMount: true })
    const value = wrapper.find('.value')
    void act(() => { value?.simulate('click') })
    expect(wrapper).toMatchSnapshot()
  })

  it('should hide list', () => {
    jest.spyOn(React, 'useState').mockImplementationOnce(init => [init, useStateMock])

    const wrapper = setup({ component, props, shouldMount: true })
    const firstItem = wrapper.find('li').first()

    void act(() => { firstItem.simulate('click') })

    expect(useStateMock).toHaveBeenCalledWith(false)
  })

  it('should select item', () => {
    const wrapper = setup({ component, props, shouldMount: true })

    const firstItem = wrapper.find('li').first()
    void act(() => { firstItem.simulate('click') })

    expect(onChange).toHaveBeenCalledWith(O.some(value))
  })
})
