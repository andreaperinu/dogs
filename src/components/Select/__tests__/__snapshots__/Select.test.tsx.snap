// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should match the snapshot (blank value) 1`] = `
<styled.div
  disabled={false}
  showOptions={false}
>
  <div
    className="wrapper"
  >
    <div
      className="value"
      onClick={[Function]}
    />
    <ul
      className="list"
    >
      <li
        key="0"
        onClick={[Function]}
        value="0"
      >
        name_0
      </li>
      <li
        key="1"
        onClick={[Function]}
        value="1"
      >
        name_1
      </li>
    </ul>
  </div>
</styled.div>
`;

exports[`<Select /> should match the snapshot (show options) 1`] = `
<BaseSelect
  maybeOptions={
    Object {
      "_tag": "Some",
      "value": Array [
        Object {
          "id": "0",
          "value": "name_0",
        },
        Object {
          "id": "1",
          "value": "name_1",
        },
      ],
    }
  }
  maybeValue={
    Object {
      "_tag": "Some",
      "value": Object {
        "id": "0",
        "value": "name_0",
      },
    }
  }
  onChange={[MockFunction]}
>
  <styled.div
    disabled={false}
    showOptions={false}
  >
    <div
      className="sc-bdfBwQ dePLvY"
      disabled={false}
    >
      <div
        className="wrapper"
      >
        <div
          className="value"
          onClick={[Function]}
        >
          name_0
        </div>
        <ul
          className="list"
        >
          <li
            key="0"
            onClick={[Function]}
            value="0"
          >
            name_0
          </li>
          <li
            key="1"
            onClick={[Function]}
            value="1"
          >
            name_1
          </li>
        </ul>
      </div>
    </div>
  </styled.div>
</BaseSelect>
`;

exports[`<Select /> should match the snapshot (show text) 1`] = `
<styled.div
  disabled={false}
  showOptions={false}
>
  <div
    className="wrapper"
  >
    <div
      className="value"
      onClick={[Function]}
    >
      name_0
    </div>
    <ul
      className="list"
    >
      <li
        key="0"
        onClick={[Function]}
        value="0"
      >
        name_0
      </li>
      <li
        key="1"
        onClick={[Function]}
        value="1"
      >
        name_1
      </li>
    </ul>
  </div>
  <div
    className="text"
    onClick={[Function]}
  >
    text
  </div>
</styled.div>
`;

exports[`<Select /> should match the snapshot 1`] = `
<styled.div
  disabled={false}
  showOptions={false}
>
  <div
    className="wrapper"
  >
    <div
      className="value"
      onClick={[Function]}
    >
      name_0
    </div>
    <ul
      className="list"
    >
      <li
        key="0"
        onClick={[Function]}
        value="0"
      >
        name_0
      </li>
      <li
        key="1"
        onClick={[Function]}
        value="1"
      >
        name_1
      </li>
    </ul>
  </div>
</styled.div>
`;
